include/emp/Evolve/Systematics.hpp

```diff
@@ -1810,31 +1781,645 @@ namespace emp {
+  int Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetBranchesToRoot(Ptr<taxon_t> tax) const {
+    GetMRCA();
+
+    int depth = 0;
+    Ptr<taxon_t> test_taxon = tax->GetParent();
+    while (test_taxon) {
+      if (test_taxon == mrca || !test_taxon) {
+        return depth;
+      } else if (test_taxon->GetNumOff() > 1) {
+        depth++;
+      }
+      test_taxon = test_taxon->GetParent();
+    }
+    return depth;
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  int Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetDistanceToRoot(Ptr<taxon_t> tax) const {
+    // Now, trace the line of descent, updating the candidate as we go.
+    GetMRCA();
+
+    int depth = 0;
+    Ptr<taxon_t> test_taxon = tax->GetParent();
+    while (test_taxon) {
+      depth++;
+      if (test_taxon == mrca || !test_taxon) {
+        return depth;
+      }
+      test_taxon = test_taxon->GetParent();
+    }
+    return depth;
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  std::set<Ptr<typename Systematics<ORG, ORG_INFO, DATA_STRUCT>::taxon_t>> Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetCanopyExtantRoots(int time_point) const {
+    // NOTE: This could be made faster by doing something similar to the pairwise distance
+    // function
+    using taxon_t = Systematics<ORG, ORG_INFO, DATA_STRUCT>::taxon_t;
+    std::set< Ptr<taxon_t>> result;
+    // std::cout << "starting " << time_point << std::endl;
+    for (Ptr<taxon_t> tax : active_taxa) {
+        // std::cout << tax->GetInfo() << std::endl;
+      while (tax) {
+        // std::cout << tax->GetInfo() << " " << tax->GetOriginationTime() << " " << tax->GetDestructionTime() << std::endl;
+        if (tax->GetOriginationTime() <= time_point && tax->GetDestructionTime() > time_point ) {
+          result.insert(tax);
+        // std::cout << "inserting " << tax->GetInfo() << std::endl;
+          break;
+        }
+        tax = tax->GetParent();
+      }
+    }
+
+    return result;
+
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  int Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetPhylogeneticDiversityNormalize(int generation, std::string filename) const {
+    int gen_value = ((generation / 10) - 1); //indexes from 0, 100 generations would correspond to the 10th line in the csv
+    // bool percent_found = false;
+    int phylogenetic_diversity = ancestor_taxa.size() + active_taxa.size() - 1;
+
+    if(filename == ""){
+      //std::cout << "Phylogenetic Diversity is " << phylogenetic_diversity << std::endl;
+      return phylogenetic_diversity;
+    } else{
+
+      emp::File generation_percentiles(filename); //opens file
+      emp::vector< emp::vector<double> >percentile_data = generation_percentiles.ToData<double>(","); //turns file contents into vector
+
+      for(int j = 0; j <= percentile_data[gen_value].size() - 2; j++){ //searches through vector for slot where phylo diversity fits
```

Comment:
I don't think either of those will work because we're not looking for a value in the vector, we're looking for a place where a value would fit.

This could definitely be more efficient, but since it's such an early methodological prototype, I don't know if we need to worry a ton about that right now